### PR TITLE
Allow update absent nullable fields

### LINKED ISSUE #3378: update: allow update absent nullable fields
Now in tarantool a one can do this:
```Lua
box.cfg{}
format = {}
format[1] = {name = 'field1', type = 'unsigned'}
format[2] = {name = 'field2', type = 'unsigned'}
format[3] = {name = 'field3', type = 'unsigned', is_nullable = true}
format[4] = {name = 'field4', type = 'unsigned', is_nullable = true}
s = box.schema.create_space('test', {format = format})
pk = s:create_index('pk')
s:replace{1, 2}
```
In the inserted tuple `{1, 2}` fields 3 and 4 are 'virtual' nulls - they are interpreted as `NULL`, but are not stored. But the fields presence illusion is dispelled when I do this:
```Lua
s:update({1}, {{'!', 4, 0}})
```
It leads to an error, because update operation `!` can not insert with gaps. But lets allow to do this, if gaps are virtual `NULL`s. This operation must fill gaps with `NULL`s.

Same about `=`. Operation `#` must be able to become no-op, when a field is a virtual NULL.

### PR DESCRIPTION
**Pull request by mary3000**
Update operations can't insert with gaps.
But there exist "virtual nulls" in tarantool -
i.e. fields that are specified in format and marked as is_nullable.
They are not actually stored in the database, and currently
tarantool refuses to do update ops with such nulls as "gaps".

But from the user side these nulls exist. So this patch changes
behaviour to such that update ops will fill virtual nulls with
the real ones.

Closes #3378

### CODE REVIEW

**Comment by Gerold103** [comment]
Hi! I didn't review it in details yet, but I see you can simplify and speed up `xrow_update_op_append_nulls()` a lot. `MP_NIL` is a single byte, so you can insert a big rope batch in the end (no need to store the nils in individual rope nodes I think), and `memset()` it with the nil byte code.

**Comment by mary3000** [comment]
> Hi! I didn't review it in details yet, but I see you can simplify and speed up `xrow_update_op_append_nulls()` a lot. `MP_NIL` is a single byte, so you can insert a big rope batch in the end (no need to store the nils in individual rope nodes I think), and `memset()` it with the nil byte code.

Ok, fixed

File: src/box/xrow_update_array.c [hunk 1]
```diff
@@ -310,6 +310,36 @@ xrow_update_array_store(struct xrow_update_field *field,
...
+	uint32_t size = xrow_update_rope_size(rope);
+	if (op->field_no < 0 || (uint32_t)op->field_no <= size)
+		return 0;
+	uint32_t nil_count = op->field_no - size;
+	struct xrow_update_array_item *item =
+		(struct xrow_update_array_item *)
+		xrow_update_alloc(rope->ctx, sizeof(*item));
+	if (item == NULL)
+		return -1;
+	uint32_t nil_size = mp_sizeof_nil();
```
Comment (Gerold103) [inline src/box/xrow_update_array.c:331]: Please, add an assertion that `mp_sizeof_nil() == 1`, and drop `nil_size`.

File: src/box/xrow_update_array.c [hunk 2]
```diff
@@ -310,6 +310,36 @@ xrow_update_array_store(struct xrow_update_field *field,
...
+	uint32_t byte_count = nil_size * nil_count;
+	char *pos = (char *) region_alloc(rope->ctx, byte_count);
```
Comment (Gerold103) [inline src/box/xrow_update_array.c:333]: `Region_alloc` does not set diag. You need to do it manually.

File: src/box/xrow_update_array.c [hunk 3]
```diff
@@ -310,6 +310,36 @@ xrow_update_array_store(struct xrow_update_field *field,
...
+	if (pos == NULL)
+		return -1;
+	memset(pos, 0xc0, byte_count);
+	xrow_update_array_item_create(item, XUPDATE_NOP, pos, byte_count, 0);
+	if (xrow_update_rope_insert(rope, op->field_no, item, nil_count) != 0)
```
Comment (Gerold103) [inline src/box/xrow_update_array.c:338]: You can do `return xrow_update_rope_insert()` instead of the `if`. It might even activate the tail-call optimization.

File: test/engine/upsert.result [hunk 4]
```diff
@@ -1585,21 +1585,29 @@ s:upsert({0}, {{'!', 42, 42}})
...
+    null, null, null, null, null, null, null, null, null, null, null, null, null,
+    null, null, null, null, null, null, null, null, null, null, null, null, null]
 ...
 s:upsert({0}, {{'=', 42, 42}})
 ---
 ...
 s:select{}
 ---
-- - [0, 1, 2]
+- - [0, 1, 2, null, null, null, null, null, null, null, null, null, null, null, null,
```
Comment (Gerold103) [inline test/engine/upsert.result:1607]: I think the purpose of the test with setting a value with an index out of range was to ensure it does not work. Now it works, so better just drop these `upsert()` with 42th field.

File: test/engine/update.test.lua [hunk 5]
```diff
@@ -160,3 +160,78 @@ t:update({{'+', '[1]', 50}})
...
+s:update({1}, {{'=', -1, 128}})
+s:drop()
+
+-- '#' doesn't trim nulls
+s = box.schema.create_space('test')
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'!', 4, 0}})
+s:update({1}, {{'#', 4, 1}})
+s:drop()
```
Comment (Gerold103) [inline test/engine/update.test.lua:237]: Please, add more tests with json paths. What if I use `!` and `=` on a field whose parent's index is out of range? For example `[5][1] = 100`, where field `[5]` does not exist. The same with maps (`[5].key`).

**Review by Gerold103 — COMMENTED** [review]
Hi! Thanks for working on this!

Please, add a docbot request and a changelog file.

File: src/box/xrow_update_array.c [hunk 6]
```diff
@@ -310,6 +310,36 @@ xrow_update_array_store(struct xrow_update_field *field,
...
+	if (op->field_no < 0 || (uint32_t)op->field_no <= size)
+		return 0;
+	uint32_t nil_count = op->field_no - size;
+	struct xrow_update_array_item *item =
+		(struct xrow_update_array_item *)
+		xrow_update_alloc(rope->ctx, sizeof(*item));
+	if (item == NULL)
+		return -1;
+	assert(mp_sizeof_nil() == 1);
+	char *item_data = (char *) region_alloc(rope->ctx, nil_count);
```
Comment (Gerold103) [inline src/box/xrow_update_array.c:332]: In the new code for unary operators (including cast) we don't use whitespace after them.

File: src/box/xrow_update_array.c [hunk 7]
```diff
@@ -310,6 +310,36 @@ xrow_update_array_store(struct xrow_update_field *field,
...
+	if (item_data == NULL) {
+		diag_set(OutOfMemory, nil_count, "region", "item_data");
+		return -1;
+	}
+	memset(item_data, 0xc0, nil_count);
+	xrow_update_array_item_create(item, XUPDATE_NOP, item_data,
+			       nil_count, 0);
```
Comment (Gerold103) [inline src/box/xrow_update_array.c:339]: Misaligned line wrap.

File: test/engine/update.result [hunk 8]
```diff
@@ -850,3 +850,287 @@ s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
...
+arr = {11, 22, {111, 222}}
+---
+...
+s:replace{1, arr}
+---
+- [1, [11, 22, [111, 222]]]
+...
+s:update({1}, {{'!', '[2][42]', 0}})
+---
+- error: Field ''[2][42]'' was not found in the tuple
```
Comment (Gerold103) [inline test/engine/update.result:1128]: That does not seem right. Field `[2]` exists and it is an array. It means there should no be a reason why it can't auto-fill the fields until `[42]`. Tuple's root array is not any different from the nested arrays. At least this was the idea behind the xrow update code.

Also I realized that probably `#` should not throw an error. Because if the feature's main purpose is to fix the 'virtual' fields assignment and `=` will work on them now, then `#` should work as well. But now it fails. For example:
```
t = box.tuple.new({1, 2, 3})
t:update({{'#', 5, 1}})
```
This fails. But if I replace it with `=` or `!` - it works. That makes `#` and `!` not symmetric. Maybe such deletion should become nop when out of range or a key wasn't found (if it deletes from a map). I am afraid you need to ask Mons about that. 2 questions in total:
1) Do we auto-fill the nested arrays or only the root array?
2) Should we make `#` nop when it is out of range? To be symmetric with the new behaviour of `!`.

**Review by Gerold103 — COMMENTED** [review]
Hi! Thanks for working on this!

You still didn't add a changelog file and didn't add a docbot request. Please, do. See more comments below.

File: test/engine/update.result (hunk 8 above)
Comment (Mons) [reply test/engine/update.result:1128]: 1. I think we should auto-fill nested structures in every case, except type collision (update array field on map)
2. Agree with you, It's ok to update `#` also

Comment (Gerold103) [reply test/engine/update.result:1128]: What do you mean as every case? For instance, I have a tuple `{1, 2, 3}`. And I make update `{{'=', '[5].key[3]', 100}}`. Should it also auto-create the intermediate array and map so it would result into this: `{1, 2, 3, nil, {key = {nil, nil, 100}}}`? Or is this an error, and we only auto-fill arrays and never create new arrays and maps on the path?

Comment (Gerold103) [reply test/engine/update.result:1128]: @Mons, see above.

Comment (Mons) [reply test/engine/update.result:1128]: This is an interesting question.
By example from Perl (called autovivification) I know, that feature may lead to unexpected results, but in most cases, it is rather handy.
But since we give an option to update the deep field and give no handy option to update such a field in case of it's absence, I consider we should support auto-creation of missing structures and raise an error only in case of conflicting types (array vs map)

**Comment by EvgenyMekhanik** [comment]
WORK IN PROGRESS!!!!

File: test/engine/update.result (hunk 8 above)
Comment (EvgenyMekhanik) [reply test/engine/update.result:1146]: According to our last conversation with Mons and Alexander, we decided to make changes regarding auto-fill nested structures in a separate task

Comment (Gerold103) [reply test/engine/update.result:1146]: What about the example I showed above about filling a non-root array? It does not need creation of the nested structures, the array exists already, it is just is not in the root. And yet it does not work now.

Comment (EvgenyMekhanik) [reply test/engine/update.result:1146]: This case will also be implemented in a separate task, according to our discussion. @Mons, @alyapunov  i am right?

Comment (Gerold103) [reply test/engine/update.result:1146]: If you are not going to implement it now, it will work 50/50. For example, it will work if I will json-update some internal array using 2 operations instead of one. The second one would turn it from a bar into a rope in the code, and the auto-fill will work like in the root. But it won't work if the update operation is just one. Which looks broken and inconsistent, when result of an operation depends on presence of other independent operations. That brings the question why is such a half-done feature even needed.

Comment (alyapunov) [reply test/engine/update.result:1146]: Let's move discussion of the issue to the issue.

File: changelogs/unreleased/allow-update-absent-nullable-fields.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+## feature/core
+
+*  Update operations can't insert with gaps. This patch changes
+   the behavior so that the update operation fills the missing
+   fields with zeros.
```
Comment (Korablev77) [inline changelogs/unreleased/allow-update-absent-nullable-fields.md:5]: Missing (gh-3378) mark at the end

File: test/engine/update.test.lua [hunk 10]
```diff
@@ -160,3 +160,110 @@ t:update({{'+', '[1]', 50}})
 s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
 
 s:drop()
+
+--
+-- gh-3378: allow update absent nullable fields
+
+-- '!'
+s = box.schema.create_space('test')
```
Comment (Korablev77) [inline test/engine/update.test.lua:168]: Since this is engine test, let's check it for both memtx and vinyl (just pass engine to create_space method)

File: test/engine/update.test.lua [hunk 11]
```diff
@@ -160,3 +160,110 @@ t:update({{'+', '[1]', 50}})
...
+pk = s:create_index('pk')
+s:replace{1, 2}
+s:update({1}, {{'!', 4, 0}})
+s:drop()
```
Comment (Korablev77) [inline test/engine/update.test.lua:172]: You don't have to drop space, just use another tuple or delete it - it should be a bit faster.

File: test/engine/update.result [hunk 12]
```diff
@@ -850,3 +850,305 @@ s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
...
+---
+...
+pk = s:create_index('pk')
+---
+...
+s:replace{1, 2, box.NULL}
+---
+- [1, 2, null]
+...
+s:update({1}, {{'+', 3, 0}})
```
Comment (Korablev77) [inline test/engine/update.result:954]: Didn't get the purpose of the test: I believe the fact that update operation involving null and actual value is checked somewhere else.

File: test/engine/update.result [hunk 13]
```diff
@@ -850,3 +850,305 @@ s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
...
+...
+s:replace{1, 2}
+---
+- [1, 2]
+...
+s:update({1}, {{'!', 4, 0}})
+---
+- [1, 2, null, 0]
+...
+s:update({1}, {{'#', 4, 1}})
```
Comment (Korablev77) [inline test/engine/update.result:1076]: What about another one 'trim' operation (so that we verify that null is actually stored in the tuple - result should be [1, 2] and not [1, null])?

File: test/engine/update.test.lua (hunk 5 above)
Comment (Korablev77) [reply test/engine/update.test.lua:226]: Agree. Even behavior isn't considered to be expected (I mean like @Gerold103  suggest to do right now), test on the subj would be nice to see.

File: changelogs/unreleased/allow-update-absent-nullable-fields.md [hunk 14]
```diff
@@ -0,0 +1,5 @@
+## feature/core
+
+*  Update operations can't insert with gaps. This patch changes
+   the behavior so that the update operation fills the missing
+   fields with zeros (gh-3378).
```
Comment (Gerold103) [inline changelogs/unreleased/allow-update-absent-nullable-fields.md:5]: Not with zeros. With nulls.

File: src/box/xrow_update_array.c [hunk 15]
```diff
@@ -393,8 +426,11 @@ xrow_update_op_do_array_delete(struct xrow_update_op *op,
 
 	struct xrow_update_rope *rope = field->array.rope;
 	uint32_t size = xrow_update_rope_size(rope);
-	if (xrow_update_op_adjust_field_no(op, size) != 0)
+	if (xrow_update_op_adjust_field_no(op, size) != 0) {
```
Comment (Gerold103) [inline src/box/xrow_update_array.c:429]: This is a separate behaviour change. Better do it in another commit with its own tests.

File: test/engine/update.result [hunk 16]
```diff
@@ -850,3 +850,285 @@ s:update({1}, {{'=', 'field3', {d = 30, e = 31, f = 32}}})
 s:drop()
 ---
 ...
+--
+-- gh-3378: allow update absent nullable fields
+-- '!'
+s = box.schema.create_space('test', {engine = engine})
```
Comment (Gerold103) [inline test/engine/update.result:856]: You don't really need spaces, indexes, or any other storage things. The updates work on tuples. So for non-formatted tuples you could simply use `t = box.tuple.new(...)` and `t:update(...)`.

**Review by Gerold103 — COMMENTED** [review]
Thanks for the patch! In the commit message you duplicated the same paragraph. Please, don't. Just keep the docbot request only.

Here is the docbot request with my 2 comments inlined.
```
    Update operations can't insert with gaps. This patch changes the
    behavior so that the update operation fills the missing fields
    with zeros.
```
Firstly, the docbot request is not a 'patch'. Secondly, it fills the fields with nulls, no zeros.
```
    For example we create space s = box.schema.create_space('s'),
    then create index for this space pk = s:create_index('pk'), and then
    insert tuple in space s:insert{1, 2}. After all of this we try to update
    this tuple s:update({1}, {{'!', 5, 6}}), in previous version this operation
    fails with ER_NO_SUCH_FIELD_NO error, and after this patch it's finished
    with success and there is [1, 2, null, null, 6] tuple in space.
```
The code snippets must be wrapped into the ` signs. Otherwise some of the parts of the code might turn on a formatting rule and the text would turn into a broken mess.

**Comment by Gerold103** [comment]
When I repeatedly said that this implementation leads to inconsistent behaviour, I meant this example:
```
t = box.tuple.new({1, 2, {11, 22}})
op1 = {'=', '[3][1]', 11}
op2 = {'=', '[3][4]', 44}

-- The operation does not do anything and "works".
t:update({op1})
---
- [1, 2, [11, 22]]
...

-- The operation does not work since the array is not in the root.
t:update({op2})
---
- error: Field ''[3][4]'' was not found in the tuple
...

-- But when I combine them - magic, both work.
t:update({op1, op2})
---
- [1, 2, [11, 22, null, 44]]
...
```
This is not because of a format, or nullable/not nullable. This is because a single update operation in the xrow update code is optimized as so called 'bar' update (see `xrow_update_bar.c`). And when there is a second operation touching the same field, it is expanded to a proper array/map. The first version of the patch didn't address that, and the current version also ignores it. The example above is not a feature, it is rather a bug. The behaviour in the example literally depends on whether I add a nop update operation. And I don't see what is so hard in fixing that by making the auto-fill with nils work properly. If you got a bar update, and you see you need nils, you just need to expand it to a proper array right away.

**Comment by EvgenyMekhanik** [comment]
> When I repeatedly said that this implementation leads to inconsistent behaviour, I meant this example:
> 
> ```
> t = box.tuple.new({1, 2, {11, 22}})
> op1 = {'=', '[3][1]', 11}
> op2 = {'=', '[3][4]', 44}
> 
> -- The operation does not do anything and "works".
> t:update({op1})
> ---
> - [1, 2, [11, 22]]
> ...
> 
> -- The operation does not work since the array is not in the root.
> t:update({op2})
> ---
> - error: Field ''[3][4]'' was not found in the tuple
> ...
> 
> -- But when I combine them - magic, both work.
> t:update({op1, op2})
> ---
> - [1, 2, [11, 22, null, 44]]
> ...
> ```
> 
> This is not because of a format, or nullable/not nullable. This is because a single update operation in the xrow update code is optimized as so called 'bar' update (see `xrow_update_bar.c`). And when there is a second operation touching the same field, it is expanded to a proper array/map. The first version of the patch didn't address that, and the current version also ignores it. The example above is not a feature, it is rather a bug. The behaviour in the example literally depends on whether I add a nop update operation. And I don't see what is so hard in fixing that by making the auto-fill with nils work properly. If you got a bar update, and you see you need nils, you just need to expand it to a proper array right away.



> When I repeatedly said that this implementation leads to inconsistent behaviour, I meant this example:
> 
> ```
> t = box.tuple.new({1, 2, {11, 22}})
> op1 = {'=', '[3][1]', 11}
> op2 = {'=', '[3][4]', 44}
> 
> -- The operation does not do anything and "works".
> t:update({op1})
> ---
> - [1, 2, [11, 22]]
> ...
> 
> -- The operation does not work since the array is not in the root.
> t:update({op2})
> ---
> - error: Field ''[3][4]'' was not found in the tuple
> ...
> 
> -- But when I combine them - magic, both work.
> t:update({op1, op2})
> ---
> - [1, 2, [11, 22, null, 44]]
> ...
> ```
> 
> This is not because of a format, or nullable/not nullable. This is because a single update operation in the xrow update code is optimized as so called 'bar' update (see `xrow_update_bar.c`). And when there is a second operation touching the same field, it is expanded to a proper array/map. The first version of the patch didn't address that, and the current version also ignores it. The example above is not a feature, it is rather a bug. The behaviour in the example literally depends on whether I add a nop update operation. And I don't see what is so hard in fixing that by making the auto-fill with nils work properly. If you got a bar update, and you see you need nils, you just need to expand it to a proper array right away.

Yes, but what we must do in this case for example:
 t = box.tuple.new({1, 2, {11, 22}})
op = {'=', '[3][3][1]', 11}
create third field with array or check that first to field are not arrays and fail?
and if we have case ?
 t = box.tuple.new({1, 2, {{11, 11}, {22, 22}}})
op = {'=', '[3][3][1]', 11}

**Comment by Gerold103** [comment]
Both examples are invalid, because it was decided by Mons, that we should not create maps and arrays along the JSON path. The only thing you can do is to add nils to an array. It does not matter what are the types of the fields in array. Only field count matters.

File: test/engine/update.result (hunk 16 above)
Comment (EvgenyMekhanik) [reply test/engine/update.result:856]: Yes, but this is is engine test, and @Korablev77 write (see his comment above) that i need to check both memtx and vinyl engines. So i need to create space with specified engine.

File: changelogs/unreleased/allow-update-absent-nullable-fields.md (hunk 14 above)
Comment (EvgenyMekhanik) [reply changelogs/unreleased/allow-update-absent-nullable-fields.md]: fixed

**Review by Korablev77 — APPROVED** [review]
I guess we should push patch-set to master as (since it solves the most of user's pain) is and consider what to do with maps/arrays in scope of next release. LGTM

**Review by alyapunov — APPROVED** [review]
LGTM.
Need to create follow-up.

File: test/engine/update.result (hunk 16 above)
Comment (Gerold103) [reply test/engine/update.result:856]: This is an engine test only because you made so. You could have a separate file, or use box/ folder. Talking of the "need to check both engines" - why? It is not related to engines anyhow as I demonstrated. But up to you.

**Comment by Gerold103** [comment]
The patch still introduces inconsistent behaviour as I showed in the example above. So I can't vote it up. But you don't need my opinion here anyway, I see there are 2 other LGTMs overriding my complaint.

**Comment by Gerold103** [comment]
Btw, this commit also introduces a bug similar to this one: https://github.com/tarantool/tarantool/issues/5105. In vinyl, if the operations in my inconsistency example are in separate upserts, the behaviour will depend on whether they were merged and applied, or applied separately. I will file a ticket later.
